### PR TITLE
Crossgen2 cross target testing

### PR DESCRIPTION
**Pull request by davidwrighton**
Add cross compilation testing for crossgen2.

The goal here is to cover cross compilation for the framework libraries, with support for testing cross compilation behavior between across the large swath of possible cross target possibilities.

This testing is added to the crossgen2 outerloop and is based on the work that @echesakovMSFT built for crossgen1 for cross target support for Linux X64 to Linux Arm. (The testing model has been tweaked to allow for general purpose cross target testing.)

Important details.

There is now a cross targeting build job, which builds (on either X64 Windows or X64 Linux) the framework dlls for a given target architecture, and captures their SHA1 hashes.

Then there is a comparison job which will run crossgen2 on an arbitrary set of platforms targetting a specific OS/Architecture pair. 

The current state is that is known that the x86 compiler does not quite produce identical output when varied between 64 and 32 bit, but other architectures are expected to be succeed

### CODE REVIEW

File: eng/pipelines/coreclr/crossgen2-outerloop.yml [hunk 1]
```diff
@@ -74,3 +103,102 @@ jobs:
...
+      targetarch: x64
+
+# test target Windows X64
+# verify that cross OS targetting works
+- template: /eng/pipelines/common/platform-matrix.yml
+  parameters:
+    jobTemplate: /eng/pipelines/coreclr/templates/crossgen2-comparison-job.yml
+    buildConfig: Release
+    platforms:
+    - OSX_x64
```
Comment (trylek) [inline eng/pipelines/coreclr/crossgen2-outerloop.yml:165]: Nit - if your intent here is to just pick the most expedient platform for building something, things have changed recently and Linux is now generally considered "fastest". Most YAML scripts are using <code>CoreClrTestBuildHost</code> as the variable representing "the fastest build platform" but I understand that your requirements are somewhat specific here as a hypothetical future switch-over of CoreClrTestBuildHost to Windows thanks to availability of fast lab HW would negate the cross-OS build assumptions of this particular job matrix.

File: eng/pipelines/coreclr/templates/crossgen2-comparison-build-job.yml [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+      - template: /eng/pipelines/common/download-artifact-step.yml
+        parameters:
+          unpackFolder: $(librariesDownloadDir)
+          cleanUnpackFolder: false
+          artifactFileName: '$(librariesBuildArtifactName)$(archiveExtension)'
+          artifactName: '$(librariesBuildArtifactName)'
+          displayName: 'live-built libraries'
+
+    # Populate Core_Root
+    - script: $(Build.SourcesDirectory)/src/tests/build$(scriptExt) $(buildConfig) $(archType) $(crossArg) generatelayoutonly
```
Comment (trylek) [inline eng/pipelines/coreclr/templates/crossgen2-comparison-build-job.yml:115]: I guess that at some point we should expand this testing to exercise CG2-compiled framework (i.e. <code>crossgen2</code> or <code>composite</code> extra command-line parameters) but that can easily wait for a follow-up change.

File: eng/pipelines/coreclr/templates/crossgen2-comparison-build-job.yml [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+      displayName: Create cross-platform crossgen baseline
+      inputs:
+        scriptSource: 'filePath'
+        scriptPath: $(Build.SourcesDirectory)/src/tests/Common/scripts/crossgen2_comparison.py
+        ${{ if ne(parameters.osGroup, 'Windows_NT') }}:
+          pythonInterpreter: /usr/bin/python3
+          arguments:
+            crossgen_framework
+            --crossgen    $(crossgen2location)
+            --dotnet      $(Build.SourcesDirectory)/dotnet.sh
```
Comment (trylek) [inline eng/pipelines/coreclr/templates/crossgen2-comparison-build-job.yml:148]: Nit - at the first glance this Windows vs. Unix bifurcation seems solvable by using a few $(dir) substitutions (expanding to "/" on Unix and to "\" on Windows); if there's more to it let's leave this for a subsequent cleanup step.

**Review by trylek — APPROVED** [review]
Looks great, thanks David for pulling this off!

File: eng/pipelines/coreclr/crossgen2-outerloop.yml (hunk 1 above)
Comment (davidwrighton) [reply eng/pipelines/coreclr/crossgen2-outerloop.yml:165]: In this case I'm specifically trying to have an OSX test here to go with the other Linux and Windows tests. The more interesting case will come when we add a bit of testing around OSX arm64, but that's for a future change.

File: eng/pipelines/coreclr/templates/crossgen2-comparison-build-job.yml (hunk 2 above)
Comment (davidwrighton) [reply eng/pipelines/coreclr/templates/crossgen2-comparison-build-job.yml:115]: I don't think we really need that. In fact, looking closer, I don't actually think we need to populate core_root in this job as I believe we don't use it.

Its more interesting to think about the crossgen2-comparison-job.yml where we do use the generated runtime to run crossgen2 itself. In that case there may be a meaningful benefit to running crossgen2 when producing the core_root. OTOH, that may just make it harder to reproduce issues found by these tests.

File: eng/pipelines/coreclr/templates/crossgen2-comparison-build-job.yml (hunk 3 above)
Comment (davidwrighton) [reply eng/pipelines/coreclr/templates/crossgen2-comparison-build-job.yml:148]: There is also the difference between running dotnet.sh and dotnet.cmd.

File: eng/pipelines/coreclr/crossgen2-outerloop.yml (hunk 1 above)
Comment (davidwrighton) [reply eng/pipelines/coreclr/crossgen2-outerloop.yml:165]: And of course, Osx has a different python environment than everything else. I'll just get rid of the OSX use here. Its not strictly needed to cover the interesting scenarios, and certainly isn't worth the pain right now. We may need to revisit when we get Osx arm64 in place.

**Pull request merged**